File: C++/tree_time_instance.cpp
#include <chrono>
#include <iostream>
#include <unistd.h>
#include <queue>

using namespace std;

class Node {
public:
    Node(Node* p) {
        parent = p;
        this->data = 0;
        this->leftChild = NULL;
        this->rightChild = NULL;
    }

    Node* parent;
    Node* leftChild;
    Node* rightChild;
    int data;

    void DFS() {
        data += 1;
        if (leftChild != NULL) { leftChild->DFS(); }
        if (rightChild != NULL) {rightChild->DFS(); }
    }
};

class Tree {
public:
    Tree(int nodeNum) {
        rootNode = new Node(NULL);
        queue<Node*> nodeQueue;
        nodeQueue.push(rootNode);
        for (int i = 0; i < nodeNum; i++) {
            Node* currentNode = nodeQueue.front();
            nodeQueue.pop();
            currentNode->leftChild = new Node(currentNode);
            currentNode->rightChild = new Node(currentNode);
            nodeQueue.push(currentNode->leftChild);
            nodeQueue.push(currentNode->rightChild);

        }
    }

    void DFS() {
        rootNode->DFS();
    }
private:
    Node* rootNode;
};

int main() {
    const int NUM_TRAILS = 50;
    const int ONE_BILLION = 1000000000;
    const int ONE_MILLION = 1000000;
    std::chrono::high_resolution_clock m_clock;
    double cumulative = 0;
    for (int t = 0; t < NUM_TRAILS; t++) {
        uint64_t startTime = std::chrono::duration_cast<std::chrono::nanoseconds>(m_clock.now().time_since_epoch()).count();
        Tree* tree = new Tree(100000);
        uint64_t endTime = std::chrono::duration_cast<std::chrono::nanoseconds>(m_clock.now().time_since_epoch()).count();

        uint64_t difference = endTime - startTime;
        cumulative = cumulative + (double)difference;
        double seconds = difference / (double) ONE_BILLION;

        cout << "Time to process: " << seconds << " seconds" << '\n';
        sleep(0.003);
    }
    double seconds = cumulative / ((double) NUM_TRAILS) / ((double) ONE_BILLION);
    cout << "Total time to process: " << seconds << " seconds" << '\n';
}

File: Java/TreeTimeInstance.java
import java.util.Scanner;
import java.util.ArrayList;

class TreeTimeInstance {
    public static void main(String[] args) {
        final int ONE_MILLION = 1000000;
        final int ONE_BILLION = 1000000000;
        final int NUM_TRIALS = 50;
        
        long cumulative = 0;
        for (int t = 0; t < NUM_TRIALS; t++) {
           
            long time = System.nanoTime();
            Tree tree = new Tree(100000);
            long finalTime = System.nanoTime();

            long difference = finalTime - time;
            cumulative += difference;
            double seconds = difference / (double) ONE_BILLION;
            System.out.printf("Time to process: %f seconds\n", seconds);
        }

        try {
            Thread.sleep(3);
        } catch (InterruptedException e) { }

        double seconds = (double) cumulative / (double) ONE_BILLION / ((float) NUM_TRIALS);
        System.out.printf("Average time to process: %f seconds\n", seconds);

    }
}

class Tree {
    class Node {
        Node leftChild;
        Node rightChild;
        Node parent;
        int data;

        Node(Node parent) {
            this.parent = parent;
            data = 0;
        }

        private void DFS() {
            data++;
            if (leftChild != null) { leftChild.DFS(); }
            if (rightChild != null) {rightChild.DFS(); }
        }
    }

    Node rootNode;

    Tree(int nodeNum) {
        rootNode = new Node(null);
        ArrayList<Node> nodeQueue = new ArrayList<Node>();
        nodeQueue.add(rootNode);
        for (int i = 0; i < nodeNum; i++) {
            Node currentNode = nodeQueue.remove(0);
            currentNode.leftChild = new Node(currentNode);
            currentNode.rightChild = new Node(currentNode);
            nodeQueue.add(currentNode.leftChild);
            nodeQueue.add(currentNode.rightChild);
        }
    }

    public void DFS() {
        rootNode.DFS();
    }
}